### ISSUE TITLE
Discussion: Re-think Virtual metadata and Complex fields

### ISSUE THREAD

**Issue by paulo-graca** [DSpace/DSpace#9736]
## Is your feature request related to a problem? Please describe.
There is a clear need for DSpace to have structured fields. Sometimes we call or referred them as hierarchical metadata or nested metadata.
«Basic Nested/hierarchical metadata  (e.g. dates on a Person's affiliation with an OrgUnit)» [(DS9 discussion topics)](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+9.0+Status)

As an example for what could be `dc.contributor.author`, where we need to represent data, structurally based on the related entities and nested data:
```json
{
      "dc.title":"Graça, Paulo",
      "person.affiliation": {
          "dc.title":"DSpaceORG",
          "orgunit.identifier": {
                "orgunit.identifier.name":"ROR",
                "orgunit.identifier.id":"123456789", 
           }
      }
}
```
In the example above, we are presenting a structured `dc.contributor.author` field, related with a Person (entity) and that person with an Organization or OrgUnit, based on an affiliation field. And also a structured field for a particular identifier of type ROR.

Another example could be a `dc.relation` field of a Publication associated with a Funded Project. Like:
```json
    "dc.title":"Funded Project title",
    "project.amount": 100000
    "project.funder": {
           "dc.title": "Funder",
            "orgunit.identifier": {
                "orgunit.identifier.name":"ROR",
                "orgunit.identifier.id":"123456789", 
           }
     }
```

## Describe the solution you'd like
DSpace has an implemented concept of **virtual metadata**. You can add fields to your item that belong to related entities, through Relationships. That feature only enable you to bring static pairs `field=value` to the related item. 
Like: 
```
dc.contributor.author = Smith, John
```

The required configuration:
```xml
    <util:map id="isAuthorOfPublicationMap">
        <entry key="dc.contributor.author" value-ref="publicationAuthor_author"/>
...
    </util:map>
    <bean class="org.dspace.content.virtual.Concatenate" id="publicationAuthor_author">
        <property name="fields">
            <util:list>
                <value>person.familyName</value>
                <value>person.givenName</value>
                <value>organization.legalName</value>
            </util:list>
        </property>
        <property name="separator">
            <value>, </value>
        </property>
        <property name="useForPlace" value="true"/>
        <property name="populateWithNameVariant" value="true"/>
    </bean>
```

We need to add a layer on top of it to enable Structured data (like a new `org.dspace.content.virtual.StructuredField`):
```xml
    <util:map id="isAuthorOfPublicationMap">
        <entry key="dc.contributor.author" value-ref="publicationAuthor_structuredAuthor"/>
...
    <bean class="org.dspace.content.virtual.StructuredField" id="publicationAuthor_structuredAuthor">
        <property name="map">
            <map>
                <entry key="dc.title" value-ref="publicationAuthor_author"/>
                <entry key="person.affiliation" value-ref="publicationAuthor_affiliation"/>
...
            </map>
    </bean>
    <bean class="org.dspace.content.virtual.Concatenate" id="publicationAuthor_author">
...
    </bean>
```

What I described so far is the approach for entities with relationships, metadata that is shown on an entity that belongs to a related entity, that depends on a relationship. But I understand that the concept of virtual metadata shouldn't rely only on entities. For instance, we could add a virtual metadata field for the` Access Status` of a Publication. The access status is a runtime costly operation to determine, based on the resource policy, if the item is in Open Access status or any other state. In the retrieval we need to constantly verify what is the value for each item, based on it's bitstreams. If this field were a Publication's virtual metadata field, the value could be somehow stored or temporarily cached, avoiding the extra processing. 
This virtual metadata field could be based on the `dspace.itemType` of an item. With this, we could extend the current data.

In order to make it possible, we need to refactor the `org.dspace.content.virtual.VirtualMetadataPopulator`, to **not rely only on the relationship**, having the current solution as a particular case of a type of virtual metadata.

```xml
    <bean class="org.dspace.content.virtual.VirtualMetadataPopulator">
        <property name="map">
            <map>
                <entry key="isAuthorOfPublication" value-ref="isAuthorOfPublicationMap"/>
                <!-- relations based on OpenAIRE guidelines -->
                <entry key="isContributorOfPublication" value-ref="isAuthorOfPublicationMap"/>
            </map>
        </property>
    </bean>
```

## Additional information
Besides Virtual Metadata, DSpace would also need to support **Complex Fields** on several other Areas like Discovery/Solr storing and Retrieving, OAI, Restful retrieving responses.

I can clearly see benefits on performance and usage boost if, for searching purposes, we rely only on Solr to retrieve the stored structured data. Without having the need to perform Postgres queries, or minimize them to the bare minimum. 
As an example, our system that has Access Status feature enabled performs a total of 856 database queries for a single user only to show two pages: The homepage with the last 5 submitted records and a search result with 10 records.

This type of structured fields could open DSpace to advanced features like be compatible with:
https://openaire-guidelines-for-cris-managers.readthedocs.io/en/v1.2.0/cerif_xml_publication_entity.html

**Comment by tdonohue** [comment]
I agree this relates back to the `"Basic Nested/hierarchical metadata (e.g. dates on a Person's affiliation with an OrgUnit)"` topic listed on the 9.0 discussion topics.  So, I'll link this ticket from that page.  (There's no guarantees this topic will be scheduled for 9.0, but this discussion ticket may still be a useful resource in prioritization)

I also want to point out here that we may want to look at how this has been done in the DSpace-CRIS product, as (if I remember correctly) they already support these complex / hierarchical fields and have compatibility with CERIF.  So, we may want to compare any new brainstorms to their existing approach before doing any major prototyping/dev work on this ticket.

**Comment by mrwerdo** [comment]
Another problem I've recently faced is virtual metadata but just calculating it based on the item's own fields. I don't think it's possible with the virtual metadata file to pull fields from the item it's self.

My use case is calculating dc.title based on other fields: `dc.title =  'Issue ' + publicationissue.issueNumber`, which makes items without a logical title much nice in the user interface.
